pallets/basic-sto/src/lib.rs

```diff
@@ -0,0 +1,174 @@
+
+use codec::{Decode, Encode};
+use frame_support::{
+    decl_error, decl_event, decl_module, decl_storage, dispatch::DispatchResult, ensure,
+};
+use frame_system::{self as system, ensure_signed};
+use pallet_identity as identity;
+use pallet_settlement::{Leg, SettlementType};
+use polymesh_common_utilities::{
+    traits::{asset::Trait as AssetTrait, identity::Trait as IdentityTrait, CommonTrait},
+    Context,
+};
+use polymesh_primitives::{IdentityId, Ticker};
+use sp_runtime::traits::{CheckedMul, Saturating};
+use sp_std::prelude::*;
+
+type Identity<T> = identity::Module<T>;
+type Settlement<T> = pallet_settlement::Module<T>;
+
+pub trait Trait:
+    frame_system::Trait + CommonTrait + IdentityTrait + pallet_settlement::Trait
+{
+    /// The overarching event type.
+    type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+}
+
+/// Details about the Fundraiser
+#[cfg_attr(feature = "std", derive(Debug))]
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Fundraiser<Balance> {
+    /// Token to raise funds in
+    pub raise_token: Ticker,
+    /// Amount of offering token available for sale
+    pub remaining_amount: Balance,
+    /// Price of one million offering token units (one full token) in terms of raise token units
+    pub price_per_token: u128,
+    /// Id of the venue to use for this fundraise
+    pub venue_id: u64,
+}
+
+decl_event!(
+    pub enum Event<T>
+    where
+        Balance = <T as CommonTrait>::Balance,
+    {
+        /// A new fundraiser has been created
+        /// (offering token, raise token, amount to sell, price, venue id, fundraiser_id)
+        FundraiserCreated(IdentityId, Ticker, Ticker, Balance, u128, u64, u64),
+        /// An investor invested in the fundraiser
+        /// (offering token, raise token, offering_token_amount, raise_token_amount, fundraiser_id)
+        FundsRaised(IdentityId, Ticker, Ticker, Balance, Balance, u64),
+    }
+);
+
+decl_error! {
+    /// Errors for the Settlement module.
+    pub enum Error for Module<T: Trait> {
+        /// Sender does not have required permissions
+        Unauthorized,
+        /// An arithmetic operation overflowed
+        Overflow,
+        /// Not enough tokens left for sale
+        InsufficientTokensRemaining
+    }
+}
+
+decl_storage! {
+    trait Store for Module<T: Trait> as StoCapped {
+        /// All fundraisers that are currently running. (ticker, fundraiser_id) -> Fundraiser
+        Fundraisers get(fn fundraisers): double_map hasher(blake2_128_concat) Ticker, hasher(twox_64_concat) u64 => Fundraiser<T::Balance>;
+        /// Total fundraisers created for a token
+        FundraiserCount get(fn fundraiser_count): map hasher(twox_64_concat) Ticker => u64;
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        type Error = Error<T>;
+
+        fn deposit_event() = default;
+
+        /// Create a new offering.
```

Comment:
Comment doesn't much more than function name; consider elaborating, and on other functions below.